sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/orc/OrcColumnarBatchReader.scala

```diff
@@ -0,0 +1,432 @@
+              val data = fromColumn.asInstanceOf[BytesColumnVector]
+              var index = 0
+              while (index < batchSize) {
+                toColumn.appendByteArray(
+                  data.vector(index), data.start(index), data.length(index))
+                index += 1
+              }
+            case BinaryType =>
+              val data = fromColumn.asInstanceOf[BytesColumnVector]
+              var index = 0
+              while (index < batchSize) {
+                toColumn.appendByteArray(data.vector(index), data.start(index), data.length(index))
+                index += 1
+              }
+
+            case DecimalType.Fixed(precision, scale) =>
+              val data = fromColumn.asInstanceOf[DecimalColumnVector]
+              var index = 0
+              while (index < batchSize) {
+                val d = data.vector(index)
+                appendDecimalWritable(toColumn, precision, scale, d)
+                index += 1
+              }
+
+            case dt =>
+              throw new UnsupportedOperationException(s"Unsupported Data Type: $dt")
+          }
+        } else {
+          for (index <- 0 until batchSize) {
+            if (fromColumn.isNull(index)) {
+              toColumn.appendNull()
+            } else {
+              field.dataType match {
+                case BooleanType =>
+                  val data = fromColumn.asInstanceOf[LongColumnVector].vector(index) == 1
+                  toColumn.appendBoolean(data)
+
+                case ByteType =>
+                  val data = fromColumn.asInstanceOf[LongColumnVector].vector(index).toByte
+                  toColumn.appendByte(data)
+                case ShortType =>
+                  val data = fromColumn.asInstanceOf[LongColumnVector].vector(index).toShort
+                  toColumn.appendShort(data)
+                case IntegerType | DateType =>
+                  val data = fromColumn.asInstanceOf[LongColumnVector].vector(index).toInt
+                  toColumn.appendInt(data)
+                case LongType =>
+                  val data = fromColumn.asInstanceOf[LongColumnVector].vector(index)
+                  toColumn.appendLong(data)
+
+                case TimestampType =>
+                  val data = fromColumn.asInstanceOf[TimestampColumnVector]
+                  toColumn.appendLong(fromTimestampColumnVector(data, index))
+
+                case FloatType =>
+                  val data = fromColumn.asInstanceOf[DoubleColumnVector].vector(index).toFloat
+                  toColumn.appendFloat(data)
+                case DoubleType =>
+                  val data = fromColumn.asInstanceOf[DoubleColumnVector].vector(index)
+                  toColumn.appendDouble(data)
+
+                case StringType =>
+                  val v = fromColumn.asInstanceOf[BytesColumnVector]
+                  toColumn.appendByteArray(v.vector(index), v.start(index), v.length(index))
+
+                case BinaryType =>
```

Comment:
Can we merge `StringType` and `BinaryType` into one case?

Reply:
Yep. I merged them.